### PR TITLE
FEATURE: Implement SiteSetting to Allow Anonymous Likes

### PR DESCRIPTION
**Pull request by meltingmettle**
This pull request adds a SiteSetting to allow admins to enable or disable anonymous liking.  

Changes:
- Add `allow_anonymous_likes` SiteSetting
- Tweak `is_my_own?` to recognize "like" post actions for anonymous users, since this behavior is required in order to allow anonymous users to remove their likes.
- Update post_serializer to remove the `:can_act` scope from anonymous users after `post_undo_action_window_mins` has closed.

### CODE REVIEW

File: spec/lib/guardian_spec.rb [hunk 1]
```diff
@@ -98,6 +98,38 @@
     fab!(:user) { Fabricate(:user) }
     fab!(:post) { Fabricate(:post) }
 
+    describe "an anonymous user" do
+      before { SiteSetting.allow_anonymous_posting = true }
```
Comment (kevinchen234) [inline spec/lib/guardian_spec.rb:102]: Hmm, I've never seen this convention before.  Wondering if you want to mock this instead?
```suggestion
      before do
        allow(SiteSetting).
          to receive(:allow_anonymous_posting).
          and_return(true)
      end
```

**Review by pmusaraj — DISMISSED** [review]
Thanks @meltingmettle, the PR looks good. I am curious, what does the like popup look like in this scenario? I would expect it to be a list of anonymized user avatars. Is that correct? 

Approving, will look into merging it shortly.

**Review by nattsw — CHANGES_REQUESTED** [review]
I think there are some UX issues here:

We still need to comply to `post_undo_action_window_mins`, which isn't happening. Ideally after X mins, the anonymous user should see this.

Screenshot 2023-06-23 at 2 08 21 PM

The current bug now is that when the anon user clicks on the red ❤️ to undo the like, the frontend shows that the "like" is removed, but when you refresh the page it is actually still liked.

File: config/site_settings.yml [hunk 2]
```diff
@@ -664,6 +664,9 @@ users:
   allow_anonymous_posting:
     default: false
     client: true
+  allow_anonymous_likes:
```
[the remark this answers is not in the document]
Comment (meltingmettle) [reply config/site_settings.yml:667]: Thank you for the review!  I've addressed your comments in the latest commits. 😀

File: lib/guardian.rb [hunk 3]
```diff
@@ -620,10 +620,14 @@ def is_me?(other)
...
 
   def is_my_own?(obj)
-    unless anonymous?
-      return obj.user_id == @user.id if obj.respond_to?(:user_id) && obj.user_id && @user.id
-      return obj.user == @user if obj.respond_to?(:user)
+    if anonymous?
+      return(
+        SiteSetting.allow_anonymous_likes? && obj.class == PostAction && obj.is_like? &&
+          obj.user_id == @user.id
+      )
```
Comment (meltingmettle) [inline lib/guardian.rb:627]: stree really wanted this. :/ 

Originally this looked like:
`return SiteSetting.allow_anonymous_likes? && obj.class == PostAction && obj.is_like? && obj.user_id == @user.id if if anonymous?`

**Comment by meltingmettle** [comment]
Friendly bump!  Thank you! @nattsw @pmusaraj

**Comment by nattsw** [comment]
Hey there @meltingmettle, the issue hasn't been addressed. This is what I am seeing now with the newest commit.

https://github.com/discourse/discourse/assets/1555215/5fc3705c-5079-4c01-9251-87ef5c9d5c2a

This is what needs to be shown:

https://github.com/discourse/discourse/assets/1555215/b4e10c52-1a54-4909-b723-191ddf97e0ac

We need to abide by `post_undo_action_window_mins`. This **disallows** the user from unliking the post after the value in the site setting.

**Comment by nattsw** [comment]
@meltingmettle 

I traced the attribute from the css class:

Screenshot 2023-07-17 at 10 45 09 PM

https://github.com/discourse/discourse/blob/a982b6765f9e2da14213f36061fe3bafec00d701/app/assets/javascripts/discourse/app/widgets/post-menu.js#L166-L181

then 

https://github.com/discourse/discourse/blob/3bde7ae4338edaffac44b8fdce9f97cb0e6f5b3b/app/assets/javascripts/discourse/app/lib/transform-post.js#L260

then 

https://github.com/discourse/discourse/blob/3bde7ae4338edaffac44b8fdce9f97cb0e6f5b3b/app/assets/javascripts/discourse/app/models/action-summary.js#L7

then, I suspect the problem is with this line:

https://github.com/discourse/discourse/blob/e214fc38ed14143e702a9e3f2d2b7a2a0ffbc572/app/serializers/post_serializer.rb#L299-L308

In the case where the user is not anon, the "can_act" value is not present. Conversely, when the user is anon, "can_act" is present. There's probably an issue here with the scope.

Can you take a stab at figuring this one?

**Comment by meltingmettle** [comment]
Thank you for the tip! That was super helpful! I’ve pushed a patch, manually verified the behavior, and added additional test coverage. 

It’s ready for another round of review when you have the chance. 😄

**Review by nattsw — APPROVED** [review]
Woohoo!

🚀 

Will merge this once tests are passing

EDIT: Merged `main` due to failing spec due to webdriver update.

**Pull request merged**